share/shaders/DVR3rdPassMode1.frag

```diff
@@ -0,0 +1,192 @@
+#version 410 core
+
+in vec4 gl_FragCoord;
+layout(location = 0) out vec4 color;
+
+uniform sampler2D  backFaceTexture;
+uniform sampler2D  frontFaceTexture;
+uniform sampler3D  volumeTexture;
+uniform usampler3D missingValueMaskTexture; // !!unsigned integer!!
+uniform sampler1D  colorMapTexture;
+
+uniform vec3  someVec3[3];
+uniform ivec3 volumeDims;        // number of vertices of this volumeTexture
+uniform ivec2 viewportDims;      // width and height of this viewport
+uniform vec4  clipPlanes[6];     // clipping planes in **un-normalized** model coordinates
+
+uniform float stepSize1D;
+uniform bool  flags[3];
+uniform float lightingCoeffs[4];
+
+uniform mat4 MV;
+uniform mat4 Projection;
+uniform mat4 inversedMV;
+
+// 
+// Derive helper variables
+//
+const float ULP        = 1.2e-7f;
+const float ULP10      = 1.2e-6f;
+bool  fast             = flags[0];                  // fast rendering mode
+bool  lighting         = fast ? false : flags[1];   // no lighting in fast mode
+bool  hasMissingValue  = flags[2];                  // has missing values or not
+float ambientCoeff     = lightingCoeffs[0];
+float diffuseCoeff     = lightingCoeffs[1];
+float specularCoeff    = lightingCoeffs[2];
+float specularExp      = lightingCoeffs[3];
+vec3  boxMin           = someVec3[0];       // min coordinates of the bounding box of this volume
+vec3  boxMax           = someVec3[1];       // max coordinates of the bounding box of this volume
+vec3  colorMapRange    = someVec3[2];
+vec3  volumeDimsf      = vec3( volumeDims );
+vec3  boxSpan          = boxMax - boxMin;
+mat4  transposedInverseMV = transpose(inversedMV);
+
+//
+// Input:  Location to be evaluated in texture coordinates and model coordinates.
+// Output: If this location should be skipped.
+// Note:   It is skipped in two cases: 1) it represents a missing value
+//                                     2) it is outside of clipping planes
+//
+bool ShouldSkip( const in vec3 tc, const in vec3 mc )
+{
+    if( hasMissingValue && (texture(missingValueMaskTexture, tc).r != 0u) )
+        return true;
+
+    vec4 positionModel = vec4( mc, 1.0 );
+    for( int i = 0; i < 6; i++ )
+    {
+        if( dot(positionModel, clipPlanes[i]) < 0.0 )
+            return true;
+    }
+
+    return false;
+}
+
+//
+// Input:  Location to be evaluated in texture coordinates
+// Output: Gradient at that location
+//
+vec3 CalculateGradient( const in vec3 tc )
```

Comment:
CalculateGradient is used by Isosurface and DVR. Please put in a separate file: don't cut-n-paste code with this many lines.

Reply:
GLSL only supports `#include` in its extension profile, but not core profile. What to do in this case?

Reply:
Ask @StasJ